### DISCUSSION TITLE
Support for Partitioned Tables in PostgreSQL

### DISCUSSION THREAD

**Discussion by faycal-inajjarane** [directus/directus#21380] [Feature Requests]
### Summary

I would like to propose the addition of support for PostgreSQL partitioned tables within Directus.io. Partitioned tables are essential for managing large datasets efficiently by dividing them into smaller, more manageable pieces. This feature would significantly enhance Directus's capabilities for high-volume data management, improving performance and scalability.

**Problem Statement**
Currently, Directus.io does not support partitioned tables in PostgreSQL. When a partitioned table is created, Directus scans and detects all partitions as separate tables but fails to recognize the parent table. This behavior makes it challenging to manage large datasets effectively, as partitioning is a critical strategy for performance optimization in database management.

![image](https://github.com/directus/directus/assets/926541/05f4b273-d1de-44fe-8973-6c740093d356)

_________________________________________

![image](https://github.com/directus/directus/assets/926541/82f126bd-593f-41a7-8b14-89551952bf51)

_________________________________________

![image](https://github.com/directus/directus/assets/926541/d659376d-e731-4167-9bc8-104db68bfd40)

### Basic Example

I suggest that Directus.io be enhanced to recognize and properly handle partitioned tables in PostgreSQL. This enhancement should include:

1. Detecting and treating the parent table as a single entity, rather than recognizing only its partitions.
2. Providing the ability to interact with the partitioned table through the Directus interface as if it were a regular table, abstracting the underlying partitioning to the end-user.
3. Ensuring that Directus's API and interface can handle data insertion, querying, and management operations transparently, regardless of the partitioning.

### Motivation

**Benefits**

- Improved Performance: By supporting partitioned tables, Directus can better handle large datasets, leading to faster query times and better overall performance.

- Enhanced Scalability: Partitioning support would make Directus more suitable for enterprise-level applications where data volume can grow significantly.

- Increased Flexibility: This feature would allow developers and database administrators to optimize their database schema for performance without sacrificing the benefits of using Directus as their CMS.

**Use Cases**

- Large-scale applications with extensive datasets that require efficient querying and data management.

- Projects that involve time-series data, where partitioning by time intervals can significantly improve performance.

- Enterprises looking to migrate to Directus without compromising on database performance optimization techniques.

### Detailed Design
**Overview**
This RFC proposes an enhancement to Directus.io to support PostgreSQL partitioned tables, enabling Directus to manage large datasets efficiently by recognizing and properly interacting with partitioned tables as unified entities. This support will facilitate better performance, scalability, and flexibility in handling extensive datasets.

**Technical Background**
Partitioned tables in PostgreSQL allow for dividing a table into smaller, more manageable pieces, called partitions, based on specific keys such as ranges of dates or values. This mechanism improves query performance on large datasets and simplifies data management tasks like data retention and archival.

Currently, Directus recognizes each partition as a separate table and does not acknowledge the parent table, leading to challenges in managing partitioned data seamlessly through the Directus interface.

### Proposed Implementation
**Parent Table Recognition:**

Modify the Directus database abstraction layer to detect partitioned tables and identify their parent-child relationships.
Ensure that the parent table is treated as the primary entity in the Directus interface, with partitions being managed implicitly in the background.

**Data Interaction Enhancements:**

Enable CRUD (Create, Read, Update, Delete) operations on the parent table through the Directus API, abstracting the partition logic from the end-user.
Implement logic to route data manipulation operations to the correct partition based on the partition key.

**UI/UX Considerations:**

Update the Directus admin interface to display partitioned tables as single entities, with an option to view partition details for database administrators.
Provide visual indicators or filters in the UI to help users understand and interact with partitioned data effectively.

**API Adjustments:**

Extend the Directus API to support querying partitioned tables with existing query parameters, ensuring that partitioning details are handled transparently.

**Configuration and Documentation:**

Offer configuration options for advanced partition management, such as creating new partitions or dropping old ones directly from the Directus interface.
Include comprehensive documentation and best practices for managing partitioned tables within Directus, covering setup, configuration, and usage scenarios.
### Examples and Use Cases
**Time-Series Data Management:**
For a table partitioned by month storing event data, Directus users can insert, query, and manage events without needing to know the specifics of the partitioning scheme. The Directus interface would allow operations on the events table, while the backend ensures data is routed to the correct partition (e.g., events_2021_01, events_2021_02, etc.).

**Large Dataset Scalability:**
An e-commerce platform storing millions of order records can partition its orders table by year. Through Directus, the team can manage these records as if they were in a single table, while enjoying the performance benefits of partitioning.

### New Terminology
Parent Table: The main table in a partitioned table setup. Directus will recognize and interact with this table for data operations.
Partition/Child Table: A subdivision of the parent table containing a subset of the data, based on the partition key.

### Technical Challenges and Considerations
Partition Key Management: Ensuring seamless routing of data to the correct partition during insertions and updates.
Query Performance: Optimizing queries across partitions to maintain performance without exposing complexity to the end-user.
UI Complexity: Balancing the need for partition management features with maintaining a user-friendly interface.

### Conclusion
This RFC outlines a detailed design for integrating PostgreSQL partitioned table support into Directus, addressing the need for efficient large dataset management. By abstracting partitioning complexity and enhancing the Directus interface and API, this feature will significantly improve scalability, performance, and user experience for Directus users managing extensive datasets.

### Must Have

- Parent Table Recognition: The ability for Directus to identify and manage partitioned tables through the parent table entity, ensuring that all data manipulation and querying are seamlessly handled without requiring user awareness of the underlying partitions.

- Transparent Data Routing: Directus must automatically route data to the correct partition based on the partition key during insert, update, and delete operations.

- Partition-Aware Querying: The Directus API needs to support querying data across partitions transparently, ensuring users can retrieve data without needing to specify partition details.

- Basic UI Integration: The Directus admin interface must display partitioned tables as single entities and allow for basic data operations without exposing partition complexity to the user.

### Should Have

- Advanced Configuration Options: Provide UI and API support for advanced partition management tasks, such as creating new partitions or dropping old ones, directly from within Directus.

- Partition Management Documentation: Comprehensive documentation and best practices guides for setting up, configuring, and managing partitioned tables in Directus.

- Performance Optimization: Ensure that the implementation of partitioned table support does not negatively impact the overall performance and responsiveness of the Directus interface and API.

### Could Have

- Partition Visualization: Enhancements to the Directus UI to visualize partition structures and data distribution across partitions for database administrators and advanced users.

- Automated Partition Management: Features that allow for automatic creation and removal of partitions based on data volume or other criteria set by the user, simplifying long-term data management.

- API Extensions for Partition Control: Additional API endpoints or parameters that allow developers to perform partition-specific operations programmatically.

### Won't Have (for this version)

- Cross-Partition Transaction Management: Support for transactions that span multiple partitions, considering the complexity and potential performance implications.

- Universal Compatibility with All Database Engines: While focusing on PostgreSQL partitioned tables, ensuring this feature works identically across other database engines with partitioning capabilities (e.g., MySQL, SQL Server) might not be feasible in the initial release.

- Real-Time Partition Analytics: Providing real-time analytics and monitoring of partition health, usage, and performance directly within the Directus interface, as this could significantly increase the scope and complexity of the initial implementation.

### Code Size and Complexity: 

Introducing partitioned table support significantly increases the codebase's size and complexity. The Directus core would need to accommodate new logic for detecting, managing, and interacting with partitioned tables, which could introduce new bugs and maintenance challenges.

### Development and Maintenance Effort: 

The development effort to implement this feature is substantial, requiring expertise in both PostgreSQL partitioning and Directus's internal architecture. Additionally, ongoing maintenance would be necessary to address any partitioning-related issues that arise, especially as PostgreSQL evolves its partitioning capabilities.

### Extension-Based Partition Support:

**Design:** 
Develop a plugin or extension that provides support for partitioned tables, rather than integrating this functionality directly into the core of Directus. This plugin could offer UI components and API enhancements for managing partitioned data.

**Impact:** 
This modular approach allows for the optional inclusion of partitioning support, reducing the impact on users who do not need this feature. However, the effectiveness and integration depth of a plugin may be limited compared to a core implementation, potentially offering a less seamless experience.

### Adoption Strategy

**Enhanced Capabilities:** 
The introduction of partitioned table support can open up new possibilities for Directus projects, especially those dealing with large datasets, improving the platform's attractiveness and utility.

**Learning Curve:** 
There may be a learning curve for developers unfamiliar with database partitioning concepts. However, this can be mitigated through education and resources.

**Community Contributions:** 
As developers adopt and experiment with partitioned tables, there's potential for increased community contributions, such as new plugins or extensions to further enhance Directus's data management capabilities.

**Migration Scripts:**

-  Develop and offer migration scripts or tools to assist in the conversion of existing tables to partitioned tables where applicable. This tool could analyze a table's schema and data volume, suggesting a partitioning strategy and generating the necessary SQL commands.

- Ensure that these migration scripts carefully handle data integrity.

**Comment by mrMasly** [comment]
I believe there is no doubt that this is destined to be implemented sooner or later.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
